text/3537-msrv-resolver.md

```diff
@@ -191,7 +191,8 @@ In solving this, we need to keep in mind
   because users are stuck on 3-20 year old versions of the language specification.
   The compatibility story is fairly strong with Rust, helping us keep
   compiler and dependency upgrades cheap.
-- Some people keep their development and production MSRVs the same while others keep them separate, like with a `Cargo.msrv.lock`
+- Some people keep their development and production MSRVs the same (henceforth referred to as "shared development / publish MSRV" workflow)
+  while others keep them separate, e.g. tracking two sets of dependencies via a `Cargo.msrv.lock` (henceforth referred to as "separate development / publish MSRV" workflow)
```

Comment:
I still don't understand what is this talking about. Is this about dev-dependencies? The situation where running crate's tests requires a fairly modern rustc due to some test-only dependency, but where the crate _builds_ on much older rustcs. That is, you can _use_ the crate with Rust 1.92.0, but you need at least Rust 1.98.0 for testing?

Reply:
This is trying to describe the difference between `Cargo.lock` tracking MSRV vs `Cargo.lock` being anything and verifying MSRV another way.

Reply:
Unsure where the gap is in the description.  Contributors developing on the project are operating with one MSRV (even if its `stable`) while those depending on the package  via the registry can have a different MSRV (`package.rust-version`).

I've even seen people take this to another level by having trying to use cargo features that they assume will be stripped, like the resolver version (and we broke them by propagating it from the workspace to the package for the `cargo install` user).   I know some people were disappointed with the difficulty in verifying their published MSRV when wanting to use newer features like workspace inheritance that are completely stripped on publish.

Reply:
It may make sense to distinguish between application- and library-development. maybe that clears it up a bit?

Reply:
Maybe I'm missing it but I see that as orthogonal.  There might be preferences for different mechanisms based on those policies but they don't prescribe specific mechanisms.

Reply:
> Contributors developing on the project are operating with one MSRV (even if its stable) while those depending on the package via the registry can have a different MSRV (package.rust-version).

MSRV means _Minimum_ Supported Rust Version. If a crate supports one set of users picking one Rust version and another set of users picking a different version, then the MSRV is whichever one of those versions is lower. A crate doesn't have multiple MSRVs just because its users pick a range of Rust versions

A key distinction between application and library development is that an application generally has a single MSRV, while it is very unlikely that every single project that depends on a popular crate will have idential MSRVs.

Reply:
Yes, I would agree with that but that isn't the core of the point I was making so I'm not sure why only that part is being called out.

Reply:
> separate development / publish MSRV

RustCrypto has an MSRV of 1.41.  Their MSRV policy is that this applies to the published package for those getting the package through the registry.  This means you can depend on their packages in a project building with 1.50.

However, when committing to the repo, 1.41 is not the MSRV.  Unsure what it is (stable?).  This means you can't do basic contribution steps, like run `cargo check`, to verify your changes using 1.50 but something much newer.

The reason I'm aware of this case is we broke that, see https://github.com/rust-lang/cargo/issues/11047

> lowest MSRV or the highest MSRV

`moka` advertises its MSRV on a per-feature basis.  Currently, they are aligned but the policy states that they can drift.  See https://docs.rs/moka/0.12.1/moka/#minimum-supported-rust-versions

These are examples of what I meant by MSRV being context dependent.  Yes, it is the minimum (lowest) supported (not just by `cargo check` but by policy) rust version.  The question is what the maintainers have scoped an MSRV to and if that leaves room for additional MSRVs for other scopes.  This is similar to semantic arguments over semver in that things can technically be a breaking change but a project's policy can say otherwise (e.g. bumping MSRV or other build requirements).